src/arraymath.jl

```diff
@@ -38,13 +34,11 @@ end
     end
 end
 
-@inline fill(val, ::SA) where {SA <: StaticArray} = _fill(val, Size(SA), SA)
-@inline fill(val, ::Type{SA}) where {SA <: StaticArray} = _fill(val, Size(SA), SA)
-@generated function _fill(val::U, ::Size{s}, ::Type{SA}) where {U, s, SA <: StaticArray}
+@inline fill(val, ::SA) where {SA <: StaticArray{<:Tuple}} = _fill(val, Size(SA), SA)
+@inline fill(val::U, ::Type{SA}) where {SA <: StaticArray} where U = fill(val, SA{U})
```

Comment:
I don't pretend to know much about this code, but is it safe to add the type parameter `U` to `SA{U}` like that? What if `SA = SVector{N,T} where {N,T}` so that `SA{U} = SVector{U,T} where T`? Then the eltype `U` would be in the `N` slot instead of the `T` slot.

Would something like
```julia
@inline fill(val::U, ::Type{SA}) where {SA <: StaticArray{S}} where {U,S} = fill(val, SA{U})
```
at least ensure that the first type parameter of `SA` is filled, so that `SA{U}` fills the second slot?

Reply:
I agree, applying `U` to an arbitrary subtype of `StaticArray` is not safe.

Reply:
Thanks for the comments! I have replaced `SA{U}` with `Base.typeintersect(SA, AbstractArray{U})`.